### PR TITLE
Pin properties dev tool update

### LINKED ISSUE #12242: [DevTools] Pinned Property
**Is your feature request related to a problem? Please describe.**
When debugging with Devtools in some circumstances you need to keep the focus on only a few properties. As it is now if the properties have different names it is necessary to scroll the datagrid to see the value.

**Describe the solution you'd like**
Adding pinned properties like VS Debug Visualizer, which shows the pinned properties at the top of the property list.

**Describe alternatives you've considered**
Problem, where to store pinned properties for restore between debug session?

**Additional context**
none

### PR DESCRIPTION
**Pull request by lhsrebel72**
What does the pull request do?
This has added a pinned variable to property view models and the functionality to use that to pin/unpin properties to the top of dev tools

What is the current behavior?
there is now a checkbox in dev tools that will allow you to pin properties to the top

What is the updated/expected behavior with this PR?
it works as intended other than one thing. It is currently using two headers: one for pinned and another for unpinned properties. I would like to only have one header but, I need to figure out how to bind the widths of each column in the separate data-grids together and have been unable to so far.

How was the solution implemented (if it's not obvious)?
I implemented a check box that is associated with an IsPinned variable on PropertyViewModel. ControlDetailsViewModel will then listen for it and add pinned properties to a separate data-grid above the original one.

Checklist
 Added unit tests (if possible)?
 Added XML documentation to any related classes?
 Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation
Breaking changes
Hopefully none, fingers crossed

Obsoletions / Deprecations
Fixed issues
Fixes https://github.com/AvaloniaUI/Avalonia/issues/12242

### CODE REVIEW

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0038058-beta`. (feed url: https://nuget-feed-all.avaloniaui.net/v3/index.json) [PRBUILDID]

**Comment by robloo** [comment]
I would put both regulat and attached pinned properties in a new group in the same DataGrid

File: src/Avalonia.Controls.DataGrid/DataGrid.cs [hunk 1]
```diff
@@ -4161,7 +4161,7 @@ void SetValidationStatus(ICellEditBinding binding)
 
             ResetValidationStatus();
 
-            if (exitEditingMode)
+            if (exitEditingMode && CurrentColumn != null)
```
Comment (workgroupengineering) [inline src/Avalonia.Controls.DataGrid/DataGrid.cs:4164]: I think this change should be included in a separate PR where the reason of change is explained.

**Review by workgroupengineering — COMMENTED** [review]
Thanks for your PR, I didn't have time to do it at the time. I don't like the solution of the two headers very much because it increases the complexity. To fix you could just add one more grouping (Pinned/Unpinned) with collapsible.

**Comment by lhsrebel72** [comment]
> Thanks for your PR, I didn't have time to do it at the time. I don't like the solution of the two headers very much because it increases the complexity. To fix you could just add one more grouping (Pinned/Unpinned) with collapsible.

Do you mean something like <Grid> <DataGrid/> <DataGrid/> <Grid/>? If so, this still shows two headers while both grids are populated. When the pinned datagrid is unpopulated it is hidden though.

**Comment by workgroupengineering** [comment]
> Or more like what is shown here? https://www.c-sharpcorner.com/uploadfile/dpatra/grouping-in-datagrid-in-wpf/
> 
> I could also just simply add another groupdescription for IsPinned but, that will not achieve the effect of having the pinned properties always visible while scrolling through non-pinned properties and users would have to scroll back to the top to see them.
> 

This

**Comment by lhsrebel72** [comment]
@workgroupengineering  @robloo  I think i've addressed all concerns with the most recent commit

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0038144-beta`. (feed url: https://nuget-feed-all.avaloniaui.net/v3/index.json) [PRBUILDID]

File: src/Avalonia.Diagnostics/Diagnostics/ViewModels/AvaloniaPropertyViewModel.cs [hunk 2]
```diff
@@ -11,6 +11,12 @@ internal class AvaloniaPropertyViewModel : PropertyViewModel
         private string _priority;
         private string _group;
         private readonly Type _propertyType;
+        private enum PinnedStatus
+        {
+            Pinned,
+            Unpinned
+        }
+        private PinnedStatus _isPinned;
```
Comment (robloo) [inline src/Avalonia.Diagnostics/Diagnostics/ViewModels/AvaloniaPropertyViewModel.cs:19]: Why define this as an enum? The "Is" name prefix implies it is a bool. If you keep it as a bool you can avoid the converter entirely, correct?

File: src/Avalonia.Diagnostics/Diagnostics/ViewModels/AvaloniaPropertyViewModel.cs [hunk 3]
```diff
@@ -33,6 +40,21 @@ public AvaloniaPropertyViewModel(AvaloniaObject o, AvaloniaProperty property)
...
+            {
+                try
+                {
+                    if (Enum.TryParse(value, out PinnedStatus pinState))
+                    {
+                        _isPinned = pinState;
+                    }
+                }
+                catch { }
+            }
```
Comment (robloo) [inline src/Avalonia.Diagnostics/Diagnostics/ViewModels/AvaloniaPropertyViewModel.cs:56]: It is bad convention to be changing the data type in the property accessors. Whatever the underlying type is really should stay exposed as the property type.

File: src/Avalonia.Diagnostics/Diagnostics/ViewModels/ClrPropertyViewModel.cs [hunk 4]
```diff
@@ -9,6 +9,12 @@ internal class ClrPropertyViewModel : PropertyViewModel
         private Type _assignedType;
         private object? _value;
         private readonly Type _propertyType;
+        private enum PinnedStatus
```
Comment (robloo) [inline src/Avalonia.Diagnostics/Diagnostics/ViewModels/ClrPropertyViewModel.cs:12]: I think this enum can go away. Still, you shouldn't have two of the exact same enum defined.

File: src/Avalonia.Diagnostics/Diagnostics/ViewModels/PropertyViewModel.cs [hunk 5]
```diff
@@ -11,6 +11,7 @@ internal abstract class PropertyViewModel : ViewModelBase
         public abstract string Name { get; }
         public abstract string Group { get; }
         public abstract Type AssignedType { get; }
+        public abstract string IsPinned { get; set; }
```
Comment (robloo) [inline src/Avalonia.Diagnostics/Diagnostics/ViewModels/PropertyViewModel.cs:14]: So... three type? Bool in ConvertBack, String here and also the Enum?... should be just one as mentioned before. Am I missing something?

Comment (lhsrebel72) [reply src/Avalonia.Diagnostics/Diagnostics/ViewModels/PropertyViewModel.cs:14]: So, when it was only a bool passing it in as a groupdescription would have the groups titled true and false rather than pinned and unpinned. 
The enum was just an attempt to avoid hardcoding the strings. 

If there is another way to have the group descriptions renamed I will implement that but, I wasn't able to find anything. Even just hiding the name of the group would be a good option if possible.

**Comment by workgroupengineering** [comment]
It's a good start, but there are some problems to solve.

The first and most annoying is the loss of the pinned properties when selecting a different control  from treeview.
Secondly, the checkboxes take up space reducing the visibility of the columns
Third the property view is not updated until a different property is selected.

> The first and most annoying is the loss of the pinned properties when selecting a different control from treeview.

Instead of adding the `IsPinned` property, you could create a HashSet like this `HashSet<(String PropertyName)> pinnedProperties` and change your converter to something like this:

```csharp
    internal class PinnedConverter : IValueConverter
    {
        public HashSet<string> PrinnedProperties { get; set; }

        public object Convert(object? value, Type targetType, object? parameter, CultureInfo culture)
        {
            var isPinned = false;
            if (value is string propertyName)
            {
                isPinned = PrinnedProperties.Contains(propertyName);
            }

            return isPinned ? "Pinned" : "Unpinned";
        }

        public object ConvertBack(object? value, Type targetType, object? parameter, CultureInfo culture)
        {
            return BindingOperations.DoNothing;
        }
    }
```
and use

```csharp

private PinnedConverter pinConverter = new();
...

    pinConverter .PrinnedProperties = pinnedProperties

   // view.GroupDescriptions.Add(new DataGridPathGroupDescription(nameof(AvaloniaPropertyViewModel.IsPinned)));
   view.GroupDescriptions.Add(new DataGridPathGroupDescription(nameof(AvaloniaPropertyViewModel.PropertyName))
     {
        Converter = pinConverter ,
     }
   );

```

> Secondly, the checkboxes take up space reducing the visibility of the columns
> Third the property view is not updated until a different property is selected.

When a datagrid row is selected, you could add a AdornerDecorator that displays a button that allows you to pin/unpin the property .

**Comment by lhsrebel72** [comment]
> It's a good start, but there are some problems to solve.
> 
> The first and most annoying is the loss of the pinned properties when selecting a different control from treeview. Secondly, the checkboxes take up space reducing the visibility of the columns Third the property view is not updated until a different property is selected.
> 
> > The first and most annoying is the loss of the pinned properties when selecting a different control from treeview.
> 
> Instead of adding the `IsPinned` property, you could create a HashSet like this `HashSet<(String PropertyName)> pinnedProperties` and change your converter to something like this:
> 
> ```cs
>     internal class PinnedConverter : IValueConverter
>     {
>         public HashSet<string> PrinnedProperties { get; set; }
> 
>         public object Convert(object? value, Type targetType, object? parameter, CultureInfo culture)
>         {
>             var isPinned = false;
>             if (value is string propertyName)
>             {
>                 isPinned = PrinnedProperties.Contains(propertyName);
>             }
> 
>             return isPinned ? "Pinned" : "Unpinned";
>         }
> 
>         public object ConvertBack(object? value, Type targetType, object? parameter, CultureInfo culture)
>         {
>             return BindingOperations.DoNothing;
>         }
>     }
> ```
> 
> and use
> 
> ```cs
> private PinnedConverter pinConverter = new();
> ...
> 
>     pinConverter .PrinnedProperties = pinnedProperties
> 
>    // view.GroupDescriptions.Add(new DataGridPathGroupDescription(nameof(AvaloniaPropertyViewModel.IsPinned)));
>    view.GroupDescriptions.Add(new DataGridPathGroupDescription(nameof(AvaloniaPropertyViewModel.PropertyName))
>      {
>         Converter = pinConverter ,
>      }
>    );
> ```
> 
> > Secondly, the checkboxes take up space reducing the visibility of the columns
> > Third the property view is not updated until a different property is selected.
> 
> When a datagrid row is selected, you could add a AdornerDecorator that displays a button that allows you to pin/unpin the property .

Those seem like great suggestions, I should have some time to get that done and another commit pushed this weekend. Thanks!

**Comment by lhsrebel72** [comment]
@workgroupengineering I'm getting that wrapped up but, have never worked with adorners before and am having trouble getting that set up. Could you recommend any documentation or a tutorial to help me get that working? Or anywhere where something similar has been done in code would be a great help as well. Thanks!

**Comment by timunie** [comment]
@lhsrebel72 I've seen it in Avalonia.Behaviors for example: https://github.com/AvaloniaUI/Avalonia.Xaml.Behaviors/tree/master/src/Avalonia.Xaml.Interactions.Draggable

**Comment by workgroupengineering** [comment]
@timunie  got ahead of me.
